cloudpickle/cloudpickle.py

```diff
@@ -460,6 +460,31 @@ def func():
                             # then discards the reference to it
                             self.write(pickle.POP)
 
+    def _save_dynamic_enum(self, obj):
+        """Special handling for dynamic Enum subclasses
+
+        Use the Enum functional API (inherited from EnumMeta.__call__) as the
+        EnumMeta metaclass has complex initialization that makes the Enum
+        subclasses hold references to their own instances.
+        """
+        # XXX: shall we pass type and start kwargs? If so how to retrieve the
+        # correct info from obj.
+        elements = dict((e.name, e.value) for e in obj)
+
+        if obj.__doc__ is not obj.__base__.__doc__:
+            doc = obj.__doc__
+        else:
+            doc = None
+
+        extra = {}
+        if hasattr(obj, "__qualname__"):
+            extra["qualname"] = obj.__qualname__
+
+        self.save_reduce(_make_dynamic_enum,
+                         (obj.__base__, obj.__name__, elements, doc,
```

Comment:
It should be `obj.__bases__[1]` I think.  `obj.__bases__[-1]` is going to be `enum.EnumMeta` in this case and this is probably not what we want here.

Reply:
> `obj.__bases__[-1]` is going to be `enum.EnumMeta`

`enum.EnumMeta` is the type of `obj`, no? I think `obj.__bases__` contains the class inherited by `obj` (like `enum.IntEnum` for instance).

> It should be `obj.__bases__[1]`

I think (athough not 100% sure) that `obj.__bases__` can only be of size 1 or 2 for `Enum` classes, to in that case that would be equivalent.

Reply:
I think we need to forward all of the bases. This list can be as long as you want, each type is a mixin added to the values of the enum. For example:

```python
In [4]: class A:
   ...:     def a(self): return 'a'
   ...:     

In [5]: class B:
   ...:     def b(self): return 'b'
   ...:     

In [6]: class ABEnum(A, B, enum.Enum):
   ...:     v1 = 1
   ...:     v2 = 2
   ...:     

In [7]: ABEnum.v2.a()
Out[7]: 'a'

In [8]: ABEnum.v2.b()
Out[8]: 'b'

In [9]: ABEnum.v2
Out[9]: <ABEnum.v2: 2>
```

The current implementation of `_make_dynamic_enum` will fail if it needs to serialize a class like `IntEnum` (note: not a subclass of `IntEnum`, `IntEnum` itself) without being able to use the `__module__` and `__name__` style class serialization. here is a failing test that demonstrates this:

```Python
    def test_locally_defined_string_enum(self):
        enum = pytest.importorskip("enum")

        class StringEnum(str, enum.Enum):
            """Enum when all members are also (and must be) strings"""

        class Color(StringEnum):
            RED = 'RED'
            GREEN = 'GREEN'
            BLUE = 'BLUE'

        green1, green2, ClonedColor = pickle_depickle(
            [Color.GREEN, Color.GREEN, Color],
            protocol=self.protocol)

        assert green1 is green2
        assert green1 is ClonedColor.GREEN
        assert green1 is not ClonedColor.BLUE
        assert ClonedColor is Color
```

`EnumMeta` does ensure that the `Enum` class part does come _last_ (https://github.com/python/cpython/blob/master/Lib/enum.py#L496-L499) so we at least know that can rely on that. I would recommend sending the entire class dict (pruning some unneeded things like non-overridden methods) and the entire `__bases__` to be reconstructed on the receiving end. Another reason to serialize the entire class body is that enums may have methods, for example a custom `__repr__`. Currently, we do not send extra methods:


```Python
    def test_interactively_defined_enum_with_method(self):
        pytest.importorskip("enum")
        code = """if __name__ == "__main__":
        from enum import Enum
        from testutils import subprocess_worker

        with subprocess_worker(protocol={protocol}) as w:

            class WithMethod(Enum):
                x = 2
                y = 3

                def method(self):
                    return self.value * 2

            def check(x1, x2, ClonedWithMethod):
                assert x1 is x2
                assert x1 is ClonedWithMethod.x
                assert x1 is not ClonedWithMethod.y

                assert x1.method() == 4

            w.run(check, WithMethod.x, WithMethod.x, WithMethod)
        """.format(protocol=self.protocol)
        assert_run_python_script(code)
```

Reply:
> I think we need to forward all the bases

I agree, however this current implementation uses `EnumMeta._create_`, which only forwards two base classes: the class on which `_create_` was called, and the `type` argument. The reason why @ogrisel ended up doing it this way (instead of using `types.new_class`) is explained here, https://github.com/cloudpipe/cloudpickle/issues/244#issuecomment-462066774.